### PR TITLE
feat(CatalogTable): truncate long description with ellipsis and tooltip

### LINKED ISSUE #4297: [catalog-frontent] : truncate description and provide hover-over in entity list in home screen
Entity descriptions are usually a brief paragraph/few lines. Current entity list shows full description text, which ends up being multiple lines for a single component, makes it visually inconsistence/asymetric. As an alternative, I suggest to trunate the text after 1/2 lines and provide a (...), which can be hover-over to show tooltip with full description.

## Feature Suggestion
Use fix length for description text in list view, optionally hover-over with tooltip for complete description.

## Possible Implementation
As an alternative, I suggest to trunate the text after 1/2 lines and provide a (...), which can be hover-over to show tooltip with full description.

### PR DESCRIPTION
**Pull request by kiranpatel11**
## Hey, I just made a Pull Request for truncating longer description with ellipsis and tooltip.

Fixes #4297 

related other PR closes #4317 

#### :heavy_check_mark: Checklist

- [x] A changeset describing the change and affected packages. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#creating-changesets))
- [x] Added storybook component
- [x] Tests for new functionality and regression tests for bug fixes
- [x] Screenshots attached (for UI changes)

Screen Shot 2021-02-03 at 5 39 23 pm

### CODE REVIEW

**Comment by kiranpatel11** [comment]
Hi @freben, @adamdmharvey
Implemented in much simpler way using [React-Text-Truncate](https://github.com/ShinyChang/React-Text-Truncate). 

Please review.

File: packages/core/src/components/OverflowTooltip/OverflowTooltip.tsx [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+  line?: TextTruncateProps['line'];
+  element?: TextTruncateProps['element'];
+};
+
+export const OverflowTooltip = (props: Props) => {
+  const [hover, setHover] = useState(false);
+
+  const handleToggled = (truncated: boolean) => {
+    const hover = truncated ? true : false;
+    setHover(hover);
```
Comment (kiranpatel11) [inline packages/core/src/components/OverflowTooltip/OverflowTooltip.tsx:34]: avoided single line expression to avoid lint error

**Review by adamdmharvey — COMMENTED** [review]
Thanks for all your effort towards this, @kiranpatel11 !

A usage question: it seems you must pass text into `<OverviewToolTip>` twice, via title and text. Given the point of the component is literally to show text, and only truncate/ellipsis with tooltip if and only if it's long, presumably passing text in once should be enough? And then inside the component, it figures out the separation of truncated text to display, vs. longer text IN the tool tip?

e.g.,

```typescript
<OverviewToolTip text="This is a really long sentence, super super long, and it should truncate in some way,
but it's really really long and a run on sentence." />
```

Do you think maybe the "interface" should/could be simplified?

**Comment by kiranpatel11** [comment]
> Do you think maybe the "interface" should/could be simplified?

Made text as default title for tooltip, it can be overriden in cases different tooltip title is needed.

File: .changeset/small-bikes-enjoy.md [hunk 2]
```diff
@@ -0,0 +1,9 @@
+---
+'@backstage/core': minor
```
Comment (Rugvip) [inline .changeset/small-bikes-enjoy.md:2]: This should be a patch change, both in core and the plugin packages.

Would also be best to split this into two changesets, one for core, and another for the two plugins that just mentions the usage of the new component

File: packages/core/src/components/OverflowTooltip/OverflowTooltip.tsx [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+  const [hover, setHover] = useState(false);
+
+  const handleToggled = (truncated: boolean) => {
+    const hoverStatus = truncated ? true : false;
+    setHover(hoverStatus);
+  };
+
+  return (
+    <Tooltip
+      title={props.title ?? (props.text as string)}
```
Comment (Rugvip) [inline packages/core/src/components/OverflowTooltip/OverflowTooltip.tsx:39]: Should we make `text` always be a string in this is needed here?

**Review by Rugvip — COMMENTED** [review]
Thanks, looks good!

Couple of smaller things, but the `yarn.lock` also has too many changes. Reset it back to the version in master and re-run `yarn install` to only get the changes needed

**Review by OrkoHunter — COMMENTED** [review]
My chrome crashed when I tried loading the `yarn.lock` changes - will try again. Something feels wrong to me, 5K lines changed for adding one library. 😅 

Edit: PR seems to have unrelated commits and changes. Should sync PR with latest master.

**Comment by kiranpatel11** [comment]
> My chrome crashed when I tried loading the `yarn.lock` changes - will try again. Something feels wrong to me, 5K lines changed for adding one library. 😅
> 
> Edit: PR seems to have unrelated commits and changes. Should sync PR with latest master.

yes, yarn.lock changes came due to merge conflict. will take latest and apply my changes

File: packages/core/src/components/OverflowTooltip/OverflowTooltip.tsx (hunk 3 above)
Comment (kiranpatel11) [reply packages/core/src/components/OverflowTooltip/OverflowTooltip.tsx:39]: @Rugvip, this is just a composite component, I exposed the propType of the underneath TextTruncate component, which is optional string, whereas title prop of the MUI Tooltip does not expect undefined.

**Review by adamdmharvey — APPROVED** [review]
🎉

**Comment by kiranpatel11** [comment]
PR checks are failing due to low code coverage, I did wrote unit tests, but RTL doesnot render component with overflow style.
any suggestion ?

```
 FAIL  src/components/OverflowTooltip/OverflowTooltip.test.tsx
  <OverflowTooltip />
    ✕ renders without ellipsis (212 ms)

  ● <OverflowTooltip /> › renders without ellipsis

    TestingLibraryElementError: Unable to find an element by: [data-testid="overflow-tooltip"]

    <body>
      <div>
        <div
          style="overflow: hidden;"
        />
      </div>
    </body>

      23 |   it('renders without ellipsis', async () => {
      24 |     await renderInTestApp(<OverflowTooltip text="short text" />);
    > 25 |     expect(screen.getByTestId('overflow-tooltip')).toBeInTheDocument();
         |                   ^
      26 |     expect(screen.getByText('short text')).toBeInTheDocument();
      27 |   });
      28 | });
```

File: .changeset/small-bikes-enjoy.md (hunk 2 above)
Comment (kiranpatel11) [reply .changeset/small-bikes-enjoy.md]: done suggested changelog changes

**Review by Rugvip — COMMENTED** [review]
We can merge without full coverage or enough delta, but would be best to add at least some simple test to make sure things aren't exploding.

Good point about the `.stories` though, does need a fix :grin:

Regarding the testing troubles, could it be due to needing to wait for things to be measured and rendered? Could try awaiting `.findBy*` instead of using `.getBy*`

**Comment by kiranpatel11** [comment]
> We can merge without full coverage or enough delta, but would be best to add at least some simple test to make sure things aren't exploding.
> 
> Good point about the `.stories` though, does need a fix 😁
> 
> Regarding the testing troubles, could it be due to needing to wait for things to be measured and rendered? Could try awaiting `.findBy*` instead of using `.getBy*`

I tried some basic testcase, but its nothing much possible in this case, it only renders <div> with style(as shown above), not even data-testId. The issue is also reported @ https://github.com/testing-library/react-testing-library/issues/751 and closed without any conclusion. I can write testcase with incorrect assertion, but I don't think that serve any other purpose than to get pass the check. Indeed, that **_.stories_** exclusion would help meet the targeted thresold.

**Comment by backstage-service** [comment]
[Tugboat](https://tugboat.qa) has finished building the preview for this pull request!

Link:
- https://pr4372-gzl3wim9qz3sernv3fq0ljyptqpwegkt.tugboat.qa


Dashboard:
- https://dashboard.tugboat.qa/602b9056f285af22888f88c2

**Comment by backstage-service** [comment]
[Tugboat](https://tugboat.qa) has finished building the preview for this pull request!

Link:
- https://pr4372-gzl3wim9qz3sernv3fq0ljyptqpwegkt.tugboat.qa


Dashboard:
- https://dashboard.tugboat.qa/602b9056f285af22888f88c2

**Comment by backstage-service** [comment]
[Tugboat](https://tugboat.qa) has finished building the preview for this pull request!

Link:
- https://pr4372-gzl3wim9qz3sernv3fq0ljyptqpwegkt.tugboat.qa


Dashboard:
- https://dashboard.tugboat.qa/602b9056f285af22888f88c2

**Review by freben — APPROVED** [review]
Thanks!

**Comment by robinjmurphy** [comment]
👋 Quick question on this. Some of our components don't have description properties, which I _think_ is resulting this error:

```
chrome-extension://fmkadmapgofadopljbjfkapdkoienihi/build/react_devtools_backend.js:2430 Warning: Failed prop type: The prop `title` is marked as required in `ForwardRef(Tooltip)`, but its value is `undefined`.
    in ForwardRef(Tooltip) (created by WithStyles(ForwardRef(Tooltip)))
    in WithStyles(ForwardRef(Tooltip)) (created by OverflowTooltip)
    in OverflowTooltip (created by MTableCell)
    in td (created by ForwardRef(TableCell))
    in ForwardRef(TableCell) (created by WithStyles(ForwardRef(TableCell)))
    in WithStyles(ForwardRef(TableCell)) (created by MTableCell)
    in MTableCell (created by MTableBodyRow)
    in tr (created by ForwardRef(TableRow))
    in ForwardRef(TableRow) (created by WithStyles(ForwardRef(TableRow)))
    in WithStyles(ForwardRef(TableRow)) (created by MTableBodyRow)
    in MTableBodyRow (created by MTableBody)
    in tbody (created by ForwardRef(TableBody))
    in ForwardRef(TableBody) (created by WithStyles(ForwardRef(TableBody)))
    in WithStyles(ForwardRef(TableBody)) (created by MTableBody)
    in MTableBody
    in Unknown (created by Droppable)
    in table (created by ForwardRef(Table))
    in ForwardRef(Table) (created by WithStyles(ForwardRef(Table)))
    in WithStyles(ForwardRef(Table)) (created by Droppable)
    in div (created by Droppable)
    in div (created by Droppable)
    in div (created by Droppable)
    in Droppable (created by ConnectFunction)
    in ConnectFunction
    in ConnectFunction (created by MaterialTable)
    in div
    in Unknown (created by WithStyles(Component))
    in WithStyles(Component) (created by MaterialTable)
    in div (created by ForwardRef(Paper))
    in ForwardRef(Paper) (created by WithStyles(ForwardRef(Paper)))
    in WithStyles(ForwardRef(Paper)) (created by Container)
    in Container (created by MaterialTable)
    in Provider (created by App)
    in App (created by ErrorBoundary)
    in ErrorBoundary (created by DragDropContext)
    in DragDropContext (created by MaterialTable)
    in MaterialTable
    in Unknown (created by WithStyles(Component))
    in WithStyles(Component) (created by Table)
    in div (created by Table)
    in Table (created by CatalogTable)
    in CatalogTable (created by CatalogPageContents)
    in div (created by CatalogPageContents)
    in article (created by Content)
    in Content (created by CatalogPageContents)
    in div (created by Page)
    in ThemeProvider (created by Page)
    in Page (created by CatalogLayout)
    in CatalogLayout (created by CatalogPageContents)
    in CatalogPageContents (created by CatalogPage)
    in EntityFilterGroupsProvider (created by CatalogPage)
    in CatalogPage (created by RoutableExtensionWrapper)
    in RoutableExtensionWrapper (created by Extension(Component))
    in Suspense (created by Extension(Component))
    in Extension(Component) (at App.tsx:74)
    in Route (at App.tsx:74)
    in div (created by styled.div)
    in styled.div (at App.tsx:73)
    in FlatRoutes (at App.tsx:70)
    in div (created by SidebarPage)
    in SidebarPage (at App.tsx:68)
    in Route (created by AppRouter)
    in Routes (created by AppRouter)
    in SignInPageWrapper (created by AppRouter)
    in Router (created by BrowserRouter)
    in BrowserRouter (created by AppRouter)
    in AppRouter (at App.tsx:67)
    in RoutingProvider (created by Provider)
    in CssBaseline (created by WithStyles(CssBaseline))
    in WithStyles(CssBaseline) (created by AppThemeProvider)
    in ThemeProvider (created by AppThemeProvider)
    in AppThemeProvider (created by Provider)
    in AppContextProvider (created by Provider)
    in ApiProvider (created by Provider)
    in Provider (at App.tsx:64)
    in App (at src/index.tsx:6)
```

I'm guessing we could avoid this by just returning an empty `span` or similar here instead of creating the tooltip component? https://github.com/backstage/backstage/blob/d7d93cc32f5d9b529c2427fd81451c965b606b4f/packages/core/src/components/OverflowTooltip/OverflowTooltip.tsx#L35

**Pull request merged**